src/network/network_coordinator.cpp

```diff
@@ -0,0 +1,227 @@
+ClientNetworkCoordinatorSocketHandler _network_coordinator_client; ///< The connection to the Game Coordinator.
+ConnectionType _network_server_connection_type = CONNECTION_TYPE_UNKNOWN; ///< What type of connection the Game Coordinator detected we are on.
+
+/** Connect to the Game Coordinator server. */
+class NetworkCoordinatorConnecter : TCPConnecter {
+public:
+	/**
+	 * Initiate the connecting.
+	 * @param address The address of the Game Coordinator server.
+	 */
+	NetworkCoordinatorConnecter(const std::string &connection_string) : TCPConnecter(connection_string, NETWORK_COORDINATOR_SERVER_PORT) {}
+
+	void OnFailure() override
+	{
+		_network_coordinator_client.connecting = false;
+		_network_coordinator_client.CloseConnection(true);
+	}
+
+	void OnConnect(SOCKET s) override
+	{
+		assert(_network_coordinator_client.sock == INVALID_SOCKET);
+
+		_network_coordinator_client.sock = s;
+		_network_coordinator_client.connecting = false;
+	}
+};
+
+bool ClientNetworkCoordinatorSocketHandler::Receive_SERVER_ERROR(Packet *p)
+{
+	NetworkCoordinatorErrorType error = (NetworkCoordinatorErrorType)p->Recv_uint8();
+	std::string detail = p->Recv_string(NETWORK_ERROR_DETAIL_LENGTH);
+
+	switch (error) {
+		case NETWORK_COORDINATOR_ERROR_UNKNOWN:
+			this->CloseConnection();
+			return false;
+
+		case NETWORK_COORDINATOR_ERROR_REGISTRATION_FAILED:
+			SetDParamStr(0, detail);
+			ShowErrorMessage(STR_NETWORK_ERROR_COORDINATOR_REGISTRATION_FAILED, STR_JUST_RAW_STRING, WL_ERROR);
+
+			/* To prevent that we constantly try to reconnect, switch to private game. */
+			_settings_client.network.server_advertise = false;
+
+			this->CloseConnection();
+			return false;
+
+		default:
+			Debug(net, 0, "Invalid error type {} received from Game Coordinator", error);
+			this->CloseConnection();
+			return false;
+	}
+}
+
+bool ClientNetworkCoordinatorSocketHandler::Receive_SERVER_REGISTER_ACK(Packet *p)
+{
+	/* Schedule sending an update. */
+	this->next_update = std::chrono::steady_clock::now();
+
+	_network_server_connection_type = (ConnectionType)p->Recv_uint8();
+
+	if (_network_server_connection_type == CONNECTION_TYPE_ISOLATED) {
+		ShowErrorMessage(STR_NETWORK_ERROR_COORDINATOR_ISOLATED, STR_NETWORK_ERROR_COORDINATOR_ISOLATED_DETAIL, WL_ERROR);
+	}
+
+	SetWindowDirty(WC_CLIENT_LIST, 0);
+
+	if (_network_dedicated) {
+		std::string connection_type;
+		switch (_network_server_connection_type) {
+			case CONNECTION_TYPE_ISOLATED: connection_type = "Remote players can't connect (your server doesn't allow remote connections)"; break;
+			case CONNECTION_TYPE_DIRECT:   connection_type = "Public"; break;
+
+			case CONNECTION_TYPE_UNKNOWN: // Never returned from Game Coordinator.
+			default: connection_type = "Unknown"; break; // Should never happen, but don't fail if it does.
+		}
+
+		Debug(net, 3, " ");
```

Comment:
Won't an empty string work equally well? Now it looks a bit weird just printing a space.
On a side note I am wondering the use of an empty line in the debug/log output; it makes logs longer and will not help as much as the lines with all dashes used below.